File: ios/GroupModule.swift
import ExpoModulesCore
import moproFFI

public class GroupModule: Module {
  private var group: Group? = nil

  public func definition() -> ModuleDefinition {
    Name("Group")

    Function("initGroup") { (members: [[UInt8]]) in
      do {
        var membersData: [Data] = []
        for member in members {
          membersData.append(Data(member))
        }
        self.group = Group(members: membersData)
      } catch {
        throw Exception(name: "GroupInitError", description: "Failed to initialize group: \(error)")
      }
    }

    Function("root") {
      guard let group = self.group else {
        throw Exception(name: "GroupNotInitialized", description: "Group not initialized")
      }
      return group.root()
    }

    Function("depth") {
      guard let group = self.group else {
        throw Exception(name: "GroupNotInitialized", description: "Group not initialized")
      }
      return group.depth()
    }

    Function("members") {
      guard let group = self.group else {
        throw Exception(name: "GroupNotInitialized", description: "Group not initialized")
      }
      return group.members()
    }

    Function("indexOf") { (member: [UInt8]) in
      guard let group = self.group else {
        throw Exception(name: "GroupNotInitialized", description: "Group not initialized")
      }
      return group.indexOf(member: Data(member))
    }

    Function("addMember") { (member: [UInt8]) throws -> Void in
      guard let group = self.group else {
        throw Exception(name: "GroupNotInitialized", description: "Group not initialized")
      }
      try group.addMember(member: Data(member))
    }

    Function("addMembers") { (members: [[UInt8]]) throws -> Void in
      guard let group = self.group else {
        throw Exception(name: "GroupNotInitialized", description: "Group not initialized")
      }
      var membersData: [Data] = []
      for member in members {
        membersData.append(Data(member))
      }
      try group.addMembers(members: membersData)
    }

    Function("updateMember") { (index: UInt32, member: [UInt8]) throws -> Void in
      guard let group = self.group else {
        throw Exception(name: "GroupNotInitialized", description: "Group not initialized")
      }
      try group.updateMember(index: index, member: Data(member))
    }

    Function("removeMember") { (index: UInt32) throws -> Void in
      guard let group = self.group else {
        throw Exception(name: "GroupNotInitialized", description: "Group not initialized")
      }
      try group.removeMember(index: index)
    }
  }
}

File: android/src/main/java/expo/modules/semaphore/GroupModule.kt
package expo.modules.semaphore

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.CodedException
import java.io.File
import uniffi.mopro.*

class GroupModule : Module() {
  private var group: Group? = null

  override fun definition() = ModuleDefinition {
    Name("Group")

    Function("initGroup") { members: List<List<Int>> ->
      try {
        val membersData = members.map { member -> member.map { it.toByte() }.toByteArray() }
        group = Group(members = membersData)

      } catch (e: Exception) {
        throw CodedException("GroupInitError", e)
      }
    }

    Function("root") {
      group?.root()
    }

    Function("depth") {
      group?.depth()?.toInt() ?: 0
    }

    Function("members") {
      group?.members() ?: emptyList<ByteArray>()
    }

    Function("indexOf") { member: List<Int> ->
      group?.indexOf(member.map { it.toByte() }.toByteArray())?.toInt()
    }

    Function("addMember") { member: List<Int> ->
      group?.addMember(member.map { it.toByte() }.toByteArray())
    }

    Function("addMembers") { members: List<List<Int>> ->
      val membersData = members.map { member -> member.map { it.toByte() }.toByteArray() }
      group?.addMembers(membersData)
    }

    Function("updateMember") { index: Int, member: List<Int> ->
      group?.updateMember(index.toUInt(), member.map { it.toByte() }.toByteArray())
    }

    Function("removeMember") { index: Int ->
      group?.removeMember(index.toUInt())
    }
  }
} 